packages/seacas/libraries/ioss/src/text_mesh/Iotm_TextMesh.C

```diff
@@ -0,0 +1,626 @@
+{
+  NodeCommunicationMap commMap(m_myProcessor, map, processors);
+  commMap.verify_map_size(communication_node_count_proc());
+  commMap.fill_map_from_data(m_data);
+}
+
+void TextMesh::element_map(int64_t block_number, Ioss::Int64Vector &map) const
+{
+  raw_element_map(block_number, map);
+}
+
+void TextMesh::element_map(int64_t block_number, Ioss::IntVector &map) const
+{
+  raw_element_map(block_number, map);
+}
+
+template <typename INT>
+void TextMesh::raw_element_map(int64_t id, std::vector<INT> &map) const
+{
+  auto iter = m_blockPartition.find(id);
+  ThrowRequireMsg(iter != m_blockPartition.end(), "Could not find block with id: " << id << " in block partition");
+
+  map.reserve(iter->second.elemIds.size());
+
+  for (const auto &elemId : iter->second.elemIds) {
+    map.push_back(elemId);
+  }
+}
+
+void TextMesh::element_map(Ioss::Int64Vector &map) const
+{
+  raw_element_map(map);
+}
+
+void TextMesh::element_map(Ioss::IntVector &map) const
+{
+  raw_element_map(map);
+}
+
+template <typename INT>
+void TextMesh::raw_element_map(std::vector<INT> &map) const
+{
+  INT count = element_count_proc();
+  map.resize(count);
+
+  for (auto iter = m_blockPartition.begin(); iter != m_blockPartition.end(); iter++) {
+    size_t offset = iter->second.offset;
+    size_t blockCount = 0;
+    for (const auto &elemId : iter->second.elemIds) {
+      map[offset + blockCount++] = elemId;
+    }
+  }
+}
+
+void TextMesh::coordinates(std::vector<double> &coord) const
+{
+  /* create global coordinates */
+  int64_t count = node_count_proc();
+  coord.resize(count * m_dimension);
+  coordinates(&coord[0]);
+}
+
+void TextMesh::coordinates(double *coord) const
+{
+  if (m_rawCoordinates.empty()) return;
+
+  /* create global coordinates */
+  auto nodes = m_data.nodes_on_proc(m_myProcessor);
+  unsigned offset = 0;
+
+  for (auto node : nodes) {
+    const std::vector<double> &coords = m_coordinates[node];
+    for (unsigned i = 0; i < coords.size(); i++) {
+      coord[offset++] = coords[i];
+    }
+  }
+}
+
+void TextMesh::coordinates(std::vector<double> &x, std::vector<double> &y, std::vector<double> &z) const
+{
+  if (m_rawCoordinates.empty()) return;
+
+  /* create global coordinates */
+  int64_t count = node_count_proc();
+  x.reserve(count);
+  y.reserve(count);
+  z.reserve(count);
+
+  auto nodes = m_data.nodes_on_proc(m_myProcessor);
+
+  for (auto node : nodes) {
+    const std::vector<double> &coords = m_coordinates[node];
+
+    x.push_back(coords[0]);
+    y.push_back(coords[1]);
+    z.push_back((m_dimension == 3) ? coords[2] : 0.0);
+  }
+}
+
+void TextMesh::coordinates(int component, std::vector<double> &xyz) const
+{
+  /* create global coordinates */
+  size_t count = node_count_proc();
+  xyz.resize(count);
+  coordinates(component, xyz.data());
+}
+
+void TextMesh::coordinates(int component, double *xyz) const
+{
+  auto nodes = m_data.nodes_on_proc(m_myProcessor);
```

Comment:
The nodes_on_proc() function actually returns a std::set object and not a reference. It does a search for the processor argument on an unordered map and if it can't find it, it returns an empty set. I suppose the TextMeshData struct could hold an empty set and return that if it could not find the processor argument....this would allow a reference to be returned. Let me see what can be done about that

Reply:
OK, that sounds good.